Mathlib/Analysis/Convex/Deriv.lean

```diff
@@ -366,3 +371,319 @@ theorem strictConcaveOn_univ_of_deriv2_neg {f : ℝ → ℝ} (hf : Continuous f)
   strictConcaveOn_of_deriv2_neg' convex_univ hf.continuousOn fun x _ => hf'' x
 #align strict_concave_on_univ_of_deriv2_neg strictConcaveOn_univ_of_deriv2_neg
+
+/-!
+## Convexity of `f` implies monotonicity of `f'`
+
+In this section we prove inequalities relating derivatives of convex functions to slopes of secant
+lines, and deduce that if `f` is convex then its derivative is monotone (and similarly for strict
+convexity / strict monotonicity).
+-/
+
+/-- A non-singleton convex set in `ℝ` has the `UniqueDiffWithinAt` property, i.e. derivatives
+within `S` are unique when they exist. -/
+lemma Convex.uniqueDiffWithinAt {S : Set ℝ} (hS : Convex ℝ S) {x y : ℝ}
+    (hx : x ∈ S) (hy : y ∈ S) (hxy : x ≠ y) :
+    UniqueDiffWithinAt ℝ S x := by
+  -- Show that `𝓝[S] x` is always bounded below by either `𝓝[<] x` or `𝓝[>] x`.
+  rcases lt_or_lt_iff_ne.mpr hxy with hxy | hxy
+  · apply (uniqueDiffWithinAt_Ioi _).mono_nhds
+    rw [nhdsWithin_le_iff, mem_nhdsWithin_Ioi_iff_exists_Ioc_subset]
+    refine ⟨y, hxy, Ioc_subset_Icc_self.trans <| hS.ordConnected.out hx hy⟩
+  · apply (uniqueDiffWithinAt_Iio _).mono_nhds
+    rw [nhdsWithin_le_iff, mem_nhdsWithin_Iio_iff_exists_Ico_subset]
+    refine ⟨y, hxy, Ico_subset_Icc_self.trans <| hS.ordConnected.out hy hx⟩
+
+namespace ConvexOn
+
+variable {S : Set ℝ} {f : ℝ → ℝ} {x y : ℝ}
+
+/-- If `f : ℝ → ℝ` is convex on `S` and right-differentiable at `x ∈ S`, then the slope of any
+secant line with left endpoint at `x` is bounded below by the right derivative of `f` at `x`. -/
+lemma right_deriv_le_slope (hfc : ConvexOn ℝ S f) (hx : x ∈ S) (hy : y ∈ S) (hxy : x < y)
+    (hfd : DifferentiableWithinAt ℝ f (Ioi x) x) :
+    derivWithin f (Ioi x) x ≤ slope f x y := by
+  have := hfd.hasDerivWithinAt
+  apply le_of_tendsto <| (hasDerivWithinAt_iff_tendsto_slope' not_mem_Ioi_self).mp this
+  simp_rw [eventually_nhdsWithin_iff, slope_def_field]
+  filter_upwards [eventually_lt_nhds hxy] with t ht (ht' : x < t)
+  refine hfc.secant_mono hx (?_ : t ∈ S) hy ht'.ne' hxy.ne' ht.le
+  exact hfc.1.ordConnected.out hx hy ⟨ht'.le, ht.le⟩
+
+/-- If `f : ℝ → ℝ` is convex on `S` and differentiable within `S` at `x`, then the slope of any
+secant line with left endpoint at `x` is bounded below by `derivWithin f S x`.
+
+This is fractionally weaker than `ConvexOn.right_deriv_le_slope` but simpler to apply under a
+`DifferentiableOn S` hypothesis. -/
+lemma derivWithin_le_slope (hfc : ConvexOn ℝ S f) (hx : x ∈ S) (hy : y ∈ S) (hxy : x < y)
+    (hfd : DifferentiableWithinAt ℝ f S x) :
+    derivWithin f S x ≤ slope f x y := by
+  suffices hfd' : HasDerivWithinAt f (derivWithin f S x) (Ioi x) x from
+    hfd'.derivWithin (uniqueDiffWithinAt_Ioi _) ▸
+    hfc.right_deriv_le_slope hx hy hxy hfd'.differentiableWithinAt
+  apply hfd.hasDerivWithinAt.mono_of_mem
+  rw [mem_nhdsWithin_Ioi_iff_exists_Ioc_subset]
+  exact ⟨y, hxy, Ioc_subset_Icc_self.trans (hfc.1.ordConnected.out hx hy)⟩
```

Comment:
I think this is the perfect example where using `Has[F]Deriv[Within]At` would make your life easier. If you split `right_deriv_le_slope` into these two lemmas:
```lean
variable {f' : ℝ}

lemma le_slope_of_hasDerivWithinAt_Ioi (hfc : ConvexOn ℝ S f) (hx : x ∈ S)
    (hy : y ∈ S) (hxy : x < y) (hfd : HasDerivWithinAt f f' (Ioi x) x) :
    f' ≤ slope f x y := by
  apply le_of_tendsto <| (hasDerivWithinAt_iff_tendsto_slope' not_mem_Ioi_self).mp hfd
  simp_rw [eventually_nhdsWithin_iff, slope_def_field]
  filter_upwards [eventually_lt_nhds hxy] with t ht (ht' : x < t)
  refine hfc.secant_mono hx (?_ : t ∈ S) hy ht'.ne' hxy.ne' ht.le
  exact hfc.1.ordConnected.out hx hy ⟨ht'.le, ht.le⟩

lemma right_deriv_le_slope (hfc : ConvexOn ℝ S f) (hx : x ∈ S) (hy : y ∈ S) (hxy : x < y)
    (hfd : DifferentiableWithinAt ℝ f (Ioi x) x) :
    derivWithin f (Ioi x) x ≤ slope f x y :=
  hfc.le_slope_of_hasDerivWithinAt_Ioi hx hy hxy hfd.hasDerivWithinAt
```
Then this simply becomes:
```suggestion
lemma le_slope_of_hasDerivWithinAt (hfc : ConvexOn ℝ S f) (hx : x ∈ S) (hy : y ∈ S) (hxy : x < y)
    (hfd : HasDerivWithinAt f f' S x) :
    f' ≤ slope f x y := by
  have : S ∈ 𝓝[>] x := sorry
  refine hfc.le_slope_of_hasDerivWithinAt_Ioi hx hy hxy (hfd.mono_of_mem this)

lemma derivWithin_le_slope (hfc : ConvexOn ℝ S f) (hx : x ∈ S) (hy : y ∈ S) (hxy : x < y)
    (hfd : DifferentiableWithinAt ℝ f S x) :
    derivWithin f S x ≤ slope f x y :=
  hfc.le_slope_of_hasDerivWithinAt hx hy hxy hfd.hasDerivWithinAt
```
Side note: the sorry in this code can reasonably be its own lemma about `OrdConnected` sets imho.

Obviously you don't necessarily have to have all variations in terms of `HasDerivWithinAt` for it to work, but I think it's usually a good thing to include both versions all the time to avoid regretting it later. Does that make sense?

Reply:
I do see where you're coming from here, but is the gain really worth the additional pain of having 48 variations of the same lemma? Having 24 variations is bad enough already (particuarly since I only actually need 2 of them for the application I care about). Duplicating everything yet again by having `HasDerivAt` versions as well as `deriv` versions (and spinning out yet more little lemmas about order-connectedness), seems like a lot of very tedious effort – both for you and me right now, and for anyone who ends up having to maintain this in future.

(FWIW, the main reason I chose the `deriv` formulations over the `HasDerivAt` formulations is that it is easier to give them intuitive and systematic names! I don't find `le_slope_of_hasDerivWithinAt` very clear, it feels like it's putting the focus on the least interesting and specific object in the statement.)

Reply:
PS. Actually, having looked at this again, maybe there is more gain than I realised: in fact, rearranging it this way allows one to get rid of the `uniqueDerivWithinAt` lemmas entirely, because `HasDeriv[Within]At.neg` doesn't need any uniqueness assumptions, unlike `deriv[Within]At.neg`. I will have a shot at reworking the code using this approach and see how it comes out.

Reply:
Just to answer some of your remarks:

- regarding too many lemmas, usually Mathlib is not very conservative on that aspect, but I understand the hassle. I haven't used the calculus library extensively for some time, but in my experience if it's too annoying to duplicate everything it's actually better to just do the `Has` version, since you can just apply it to `hfd.hasDerivWithinAt`. In an ideal world we would have some automation to generate the `deriv` lemma from the `HasDeriv` version.

- regarding maintenance, of course it's an extra burden, but in that situation I don't think it would be that much more annoying (this is hard to objectively measure in advance of course).

- regarding the naming issue, I completely agree with you, and unfortunately I don't have any good answer to that. Often, I end up betting on "people will find the `deriv` version and notice the weirdly-named `HasDeriv` version just above", but this is not ideal.